File: services/certs/mtls_mesh.go
package main

import "time"

// MeshService represents a service registered in the mTLS mesh.
type MeshService struct {
	ID              string     `json:"id"`
	TenantID        string     `json:"tenant_id"`
	Name            string     `json:"name"`
	Namespace       string     `json:"namespace"`
	Endpoint        string     `json:"endpoint"`
	CertID          string     `json:"cert_id,omitempty"`
	CertCN          string     `json:"cert_cn,omitempty"`
	CertExpiry      *time.Time `json:"cert_expiry,omitempty"`
	CertStatus      string     `json:"cert_status"`
	LastRenewedAt   *time.Time `json:"last_renewed_at,omitempty"`
	AutoRenew       bool       `json:"auto_renew"`
	RenewDaysBefore int        `json:"renew_days_before"`
	TrustAnchors    []string   `json:"trust_anchors"`
	MTLSEnabled     bool       `json:"mtls_enabled"`
	CreatedAt       time.Time  `json:"created_at"`
}

// MeshCertificate represents a certificate issued to a mesh service.
type MeshCertificate struct {
	ID           string    `json:"id"`
	TenantID     string    `json:"tenant_id"`
	ServiceID    string    `json:"service_id"`
	ServiceName  string    `json:"service_name"`
	CN           string    `json:"cn"`
	SANs         []string  `json:"sans"`
	Issuer       string    `json:"issuer"`
	NotBefore    time.Time `json:"not_before"`
	NotAfter     time.Time `json:"not_after"`
	Serial       string    `json:"serial"`
	Fingerprint  string    `json:"fingerprint"`
	KeyAlgorithm string    `json:"key_algorithm"`
	Revoked      bool      `json:"revoked"`
	CreatedAt    time.Time `json:"created_at"`
}

// TrustAnchor represents a trusted CA certificate in the mesh.
type TrustAnchor struct {
	ID        string    `json:"id"`
	TenantID  string    `json:"tenant_id"`
	Name      string    `json:"name"`
	Fingerprint string  `json:"fingerprint"`
	Subject   string    `json:"subject"`
	NotBefore time.Time `json:"not_before"`
	NotAfter  time.Time `json:"not_after"`
	CreatedAt time.Time `json:"created_at"`
}

// MeshTopologyEdge represents a verified mTLS connection between two mesh services.
type MeshTopologyEdge struct {
	TenantID        string     `json:"tenant_id"`
	FromService     string     `json:"from_service"`
	ToService       string     `json:"to_service"`
	MTLSVerified    bool       `json:"mtls_verified"`
	LastHandshakeAt *time.Time `json:"last_handshake_at,omitempty"`
}

// RegisterServiceRequest is the request body for registering a mesh service.
type RegisterServiceRequest struct {
	TenantID        string   `json:"tenant_id"`
	Name            string   `json:"name"`
	Namespace       string   `json:"namespace"`
	Endpoint        string   `json:"endpoint"`
	AutoRenew       bool     `json:"auto_renew"`
	RenewDaysBefore int      `json:"renew_days_before"`
	TrustAnchors    []string `json:"trust_anchors"`
	MTLSEnabled     bool     `json:"mtls_enabled"`
}

// AddTrustAnchorRequest is the request body for adding a trust anchor.
type AddTrustAnchorRequest struct {
	TenantID    string `json:"tenant_id"`
	Name        string `json:"name"`
	Fingerprint string `json:"fingerprint"`
	Subject     string `json:"subject"`
	NotBefore   string `json:"not_before"`
	NotAfter    string `json:"not_after"`
}

File: web/dashboard/src/lib/mtlsMesh.ts
import type { AuthSession } from "./auth";
import { serviceRequest } from "./serviceApi";

export interface MeshService {
  id: string;
  name: string;
  namespace: string;
  endpoint: string;
  cert_id?: string;
  cert_cn?: string;
  cert_expiry?: string;
  cert_status: "valid" | "expiring" | "expired" | "missing";
  last_renewed_at?: string;
  auto_renew: boolean;
  renew_days_before: number;
  trust_anchors: string[];
  created_at: string;
  mtls_enabled: boolean;
}

export interface MeshCertificate {
  id: string;
  service_id: string;
  service_name: string;
  cn: string;
  san: string[];
  issuer: string;
  not_before: string;
  not_after: string;
  serial: string;
  fingerprint: string;
  key_algorithm: string;
  revoked: boolean;
  created_at: string;
}

export interface TrustAnchor {
  id: string;
  name: string;
  fingerprint: string;
  subject: string;
  not_before: string;
  not_after: string;
  services_count: number;
  created_at: string;
}

export interface MeshTopologyEdge {
  from_service: string;
  to_service: string;
  mtls_verified: boolean;
  last_handshake_at?: string;
}

export async function listServices(session: AuthSession): Promise<MeshService[]> {
  const res = await serviceRequest<any>(session, "certs", "/mesh/services");
  return res.items ?? [];
}

export async function registerService(session: AuthSession, data: Partial<MeshService>): Promise<MeshService> {
  return serviceRequest<MeshService>(session, "certs", "/mesh/services", { method: "POST", body: JSON.stringify(data) });
}

export async function renewServiceCert(session: AuthSession, serviceId: string): Promise<MeshCertificate> {
  return serviceRequest<MeshCertificate>(session, "certs", `/mesh/services/${serviceId}/renew`, { method: "POST" });
}

export async function listCertificates(session: AuthSession): Promise<MeshCertificate[]> {
  const res = await serviceRequest<any>(session, "certs", "/mesh/certificates");
  return res.items ?? [];
}

export async function listTrustAnchors(session: AuthSession): Promise<TrustAnchor[]> {
  const res = await serviceRequest<any>(session, "certs", "/mesh/trust-anchors");
  return res.items ?? [];
}

export async function getTopology(session: AuthSession): Promise<MeshTopologyEdge[]> {
  return serviceRequest<MeshTopologyEdge[]>(session, "certs", "/mesh/topology");
}
